spark2/src/main/java/org/apache/iceberg/spark/source/StreamingReader.java

```diff
@@ -0,0 +1,370 @@
+    List<MicroBatch> batches = Lists.newArrayList();
+    long currentLeftSize = maxSize;
+    MicroBatch lastBatch = null;
+
+    assertNoOverwrite(table.snapshot(offset.snapshotId()));
+    if (shouldGenerateFromStartOffset(offset)) {
+      MicroBatch batch = generateMicroBatch(offset.snapshotId(), offset.position(),
+          offset.shouldScanAllFiles(), currentLeftSize);
+      if (!batch.tasks().isEmpty()) {
+        batches.add(batch);
+        currentLeftSize -= batch.sizeInBytes();
+        lastBatch = Iterables.getLast(batches);
+      }
+    }
+
+    // Current snapshot can already satisfy the size needs.
+    if (currentLeftSize <= 0L || (lastBatch != null && !lastBatch.lastIndexOfSnapshot())) {
+      return batches;
+    }
+
+    long currentSnapshotId = table.currentSnapshot().snapshotId();
+    if (currentSnapshotId == offset.snapshotId()) {
+      // the snapshot of current offset is already the latest snapshot of this table.
+      return batches;
+    }
+
+    ImmutableList<Long> snapshotIds = ImmutableList.<Long>builder()
+        .addAll(SnapshotUtil.snapshotIdsBetween(table, offset.snapshotId(), currentSnapshotId))
+        .build()
+        .reverse();
+
+    for (Long id : snapshotIds) {
+      Snapshot snapshot = table.snapshot(id);
+      assertNoOverwrite(snapshot);
+      if (!isAppend(snapshot)) {
+        continue;
+      }
+
+      MicroBatch batch = generateMicroBatch(id, 0, false, currentLeftSize);
+      if (!batch.tasks().isEmpty()) {
+        batches.add(batch);
+        currentLeftSize -= batch.sizeInBytes();
+        lastBatch = Iterables.getLast(batches);
+      }
+
+      // If the current request size is already satisfied, or none of the DataFile size can satisfy the current left
+      // size, break the current loop.
+      if (currentLeftSize <= 0L || !Objects.requireNonNull(lastBatch).lastIndexOfSnapshot()) {
+        break;
+      }
+    }
+
+    return batches;
+  }
+
+  private MicroBatch generateMicroBatch(long snapshotId, long startIndex, boolean isStart, long currentLeftSize) {
+    return MicroBatches.from(table.snapshot(snapshotId), table.io())
+        .caseSensitive(caseSensitive())
+        .specsById(table.specs())
+        .generate(startIndex, currentLeftSize, isStart);
+  }
+
+  @SuppressWarnings("checkstyle:HiddenField")
+  private boolean shouldGenerateFromStartOffset(StreamingOffset startOffset) {
+    boolean isSnapshotFullyProcessed;
+    if (cachedPendingBatches != null && !cachedPendingBatches.second().isEmpty()) {
+      List<MicroBatch> batches = cachedPendingBatches.second();
+      MicroBatch lastBatch = Iterables.getLast(batches, null);
+      isSnapshotFullyProcessed = lastBatch.lastIndexOfSnapshot();
+    } else {
+      isSnapshotFullyProcessed = false;
+    }
+    return !isSnapshotFullyProcessed &&
+        (startOffset.shouldScanAllFiles() || isAppend(table.snapshot(startOffset.snapshotId())));
+  }
+
+  private static void assertNoOverwrite(Snapshot snapshot) {
```

Comment:
can we pl. take a configuration whether or not to assert. essentially - we want to factor in cases like compaction - which will perform harmless `REPLACE` operations.

https://docs.microsoft.com/en-us/azure/databricks/delta/delta-streaming#ignore-updates-and-deletes